File: python/Algos/reverse_linkedlist.py
class Node:
    """
    Node object for a linked list
    """
    def __init__(self, v: int):
        """
        Constructor for Linked List Node
        """
        self.val = v
        self.nxt = None

    def print(self):
        """
        Method to print all nodes in the linked list
        """
        print(f"{self.val}", end="")
        _next = self.nxt

        # when next is a nullptr, we have 
        # reached the end of the linked list
        # and this loop will stop
        while _next is not None:
            print(f"->{_next.val}", end="")
            _next = _next.nxt

        print("->NULL")

def reverse_llist(n: Node) -> Node:
    """
    Reverses a Linked List

    @param n: 1rst node of linked list
    @type n: Node

    @return The 1rst node of the reversed linked list 
    """

    prev = None
    curr = n
    _nxt = n.nxt

    # when next is a nullptr, we have 
    # reached the end of the linked list
    # and this loop will stop
    while _nxt is not None:
        curr.nxt = prev
        prev = curr
        curr = _nxt
        _nxt = _nxt.nxt

    curr.nxt = prev

    return curr


llist = Node(1)
llist.nxt = Node(2)
llist.nxt.nxt = Node(3)
llist.nxt.nxt.nxt = Node(4)
llist.nxt.nxt.nxt.nxt = Node(5)

if __name__ == "__main__":
    print("Will attempt to reverse the following Linked List:")
    llist.print()
    llist = reverse_llist(llist)
    print("Linked List after reversal:")
    llist.print()

File: go/Algos/techniques/reverse-linkedlist.go
package main

import "fmt"

/*
Node object for a linked list
*/
type Node struct {
	val  int32
	next *Node
}

/*
Method to print all nodes in the linked list
*/
func (n *Node) print() {
	fmt.Printf("%v", n.val)
	nxt := n.next

	// When next is a nullptr, we have
	// reached the end of the linked list
	// and this loop will stop
	// NOTE : Go does not contain "while" loops
	for {
		if nxt == nil {
			break
		}
		fmt.Printf("->%v", nxt.val)
		nxt = nxt.next
	}

	fmt.Println("->NULL")
}

/*
Reverses a Linked List

@param n: 1rst node of linked list
@type n: Node

@return The 1rst node of the reversed linked list
*/
func reverse_llist(n *Node) *Node {
	var prev *Node
	curr := n
	next := n.next
	for {
		if next == nil {
			break
		}
		curr.next = prev
		prev = curr
		curr = next
		next = next.next
	}

	curr.next = prev
	return curr
}

func main() {
	fmt.Println("Will attempt to reverse the following Linked List:")
	var llist *Node = &Node{1, nil}
	llist.next = &Node{2, nil}
	llist.next.next = &Node{3, nil}
	llist.next.next.next = &Node{4, nil}
	llist.next.next.next.next = &Node{5, nil}
	llist.print()
	llist = reverse_llist(llist)
	fmt.Println("Linked List after reversal:")
	llist.print()
}
